File: pro_cache_backend/test_client.js
const WebSocket = require('ws');
const fetch = require('node-fetch');

const INTERNAL_API = "http://127.0.0.1:8081/internal";
const WS_URL = "ws://127.0.0.1:8080/ws";

const TOKEN = "test-token-123";
const USER_ID = "user-1";
const PROJECT_ID = "project-A";

async function registerToken() {
    console.log(`[Client] Registering token: ${TOKEN}`);
    try {
        const response = await fetch(`${INTERNAL_API}/auth/register`, {
            method: 'POST',
            body: JSON.stringify({
                token: TOKEN,
                user_id: USER_ID,
                project_id: PROJECT_ID,
                ttl: 300
            }),
            headers: { 'Content-Type': 'application/json' }
        });
        
        if (response.ok) {
            console.log(`[Client] Register success: ${response.status}`);
            return true;
        } else {
            console.log(`[Client] Register failed: ${response.status} ${await response.text()}`);
            return false;
        }
    } catch (e) {
        console.log(`[Client] Register error: ${e}`);
        return false;
    }
}

async function triggerInvalidation() {
    await new Promise(r => setTimeout(r, 1000));
    console.log(`[Invalidator] Triggering invalidation for ${PROJECT_ID}`);
    try {
        const response = await fetch(`${INTERNAL_API}/invalidate`, {
            method: 'POST',
            body: JSON.stringify({
                project_id: PROJECT_ID,
                path: "/users",
                user_id: USER_ID
            }),
            headers: { 'Content-Type': 'application/json' }
        });
        console.log(`[Invalidator] Response: ${response.status}`);
    } catch (e) {
        console.log(`[Invalidator] Error: ${e}`);
    }
}

async function run() {
    if (!await registerToken()) process.exit(1);

    console.log("[Client] Connecting to WebSocket...");
    const ws = new WebSocket(`${WS_URL}?token=${TOKEN}`);

    ws.on('open', function open() {
        console.log("[WS] Connected");
        triggerInvalidation();
    });

    ws.on('message', function message(data) {
        console.log(`[WS] Received: ${data}`);
        // Check if it's the expected message
        if (data.includes('"/users"')) {
            console.log("[WS] Test Passed!");
            process.exit(0);
        }
    });
    
    ws.on('error', (e) => console.log(`[WS] Error: ${e}`));
}

run();

File: pro_cache_backend/test_client.py
import requests
import websocket
import json
import threading
import time
import sys

# Configuration
INTERNAL_API = "http://127.0.0.1:8081/internal"
WS_URL = "ws://127.0.0.1:8080/ws"

TOKEN = "test-token-123"
USER_ID = "user-1"
PROJECT_ID = "project-A"

def register_token():
    print(f"[Client] Registering token: {TOKEN}")
    payload = {
        "token": TOKEN,
        "user_id": USER_ID,
        "project_id": PROJECT_ID,
        "ttl": 300
    }
    try:
        resp = requests.post(f"{INTERNAL_API}/auth/register", json=payload)
        print(f"[Client] Register response: {resp.status_code} {resp.text}")
        return resp.status_code == 200
    except Exception as e:
        print(f"[Client] Register failed: {e}")
        return False

def trigger_invalidation():
    time.sleep(2) # Wait for WS connection
    print(f"[Invalidator] Triggering invalidation for {PROJECT_ID}")
    payload = {
        "project_id": PROJECT_ID,
        "path": "/users",
        "user_id": USER_ID 
    }
    try:
        resp = requests.post(f"{INTERNAL_API}/invalidate", json=payload)
        print(f"[Invalidator] Invalidate response: {resp.status_code} {resp.text}")
    except Exception as e:
        print(f"[Invalidator] Failed: {e}")

def on_message(ws, message):
    print(f"[WS] Received: {message}")
    ws.close()
    print("[WS] Test Passed!")
    sys.exit(0)

def on_error(ws, error):
    print(f"[WS] Error: {error}")

def on_close(ws, close_status_code, close_msg):
    print("[WS] Closed")

def on_open(ws):
    print("[WS] Connected")
    # Start invalidation thread
    threading.Thread(target=trigger_invalidation).start()

def connect_ws():
    # websocket.enableTrace(True)
    ws = websocket.WebSocketApp(f"{WS_URL}?token={TOKEN}",
                              on_open=on_open,
                              on_message=on_message,
                              on_error=on_error,
                              on_close=on_close)
    ws.run_forever()

if __name__ == "__main__":
    if not register_token():
        sys.exit(1)
    
    print("[Client] Connecting to WebSocket...")
    connect_ws()
